### PR TITLE
Refactored the branchlength setting and getting API.

### PR DESCRIPTION
**Pull request by Ward9250**
This is a very simple change to Phylo. It alters the branchlength and branchlength! methods so as when people create immutable types to act as branch metadata, if they just define a branchlength and branchlength! method for that type, then getting and setting the branch length for any phylogeny type should "just work".

### CODE REVIEW

File: src/phylo/branch_basics.jl [hunk 1]
```diff
@@ -10,24 +10,42 @@
...
 end
 
-function branchlength!{C}(tree::Phylogeny{C, Float64}, edge::Edge, value::Float64)
-    tree.edgedata[edge] = value
-    return tree
+# The basic function on which setting a branch length depends,
+# for any kind of phylogeny. The assumption is metadata values on branches are
+# immutables or value types - hence the reassignment.
+function branchlength!(tree::Phylogeny, edge, value::Float64)
+    tree.edgedata[edge] = branchlength!(tree.edgedata[edge], value)
```
Comment (bicycle1885) [inline src/phylo/branch_basics.jl]: Why not return `tree`?

Comment (Ward9250) [inline src/phylo/branch_basics.jl]: Sorry, it will, some stuff broke that I'm partway through altering/fixing.

**Comment by bicycle1885** [comment]
Is this ready to be merged?

**Comment by Ward9250** [comment]
Sorry no not just yet, just handed in a few days ago, just coming back to everything now and starting to remember what I was doing with this, I'll sort it tonight.

**Comment by Ward9250** [comment]
Ok, back to looking at this - I have implemented a solution I think is acceptable: for any branch metadata type you could think to annotate phylogenetic branches with, so long as you define two methods (and a third optional one in some cases), getting and setting branchlengths with `branchlength` and `branchlength!` should work for any kind of phylogeny you create with custom metadata.

**Comment by Ward9250** [comment]
I have no idea why it's included all these commits that have nothing to do with this PR.

**Comment by bicycle1885** [comment]
You can delete unnecessary commits with `git rebase -i origin/master`.

**Comment by Ward9250** [comment]
Ok tests are passing, I'd like to squash and merge this, this evening, and move subsequent Phylo development to my Fork.

**Pull request merged**